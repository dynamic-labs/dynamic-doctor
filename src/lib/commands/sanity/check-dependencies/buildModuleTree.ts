import path from 'path';
import fs from 'fs';
import { ModuleTree, PackageJson } from './types';
import { REF_SEPARATOR, ROOT_REF } from './constants';
import { isDynamicPackage } from './utils';

//////////////////////////////
// Helper Methods
//////////////////////////////

/**
 * Reads a json file and returns the parsed json.
 * @param filePath
 * @returns
 */
const readJsonFile = <T>(filePath: string) => {
  return JSON.parse(fs.readFileSync(filePath, 'utf8').trim()) as T;
};

/**
 * Merges the scoped packages into the packages array.
 * @param basePath
 * @param packages
 * @param dir
 * @returns
 */
const mergeScopes = (basePath: string, packages: string[], dir: string) => {
  if (/^@/.test(dir)) {
    const scopedSuffixPackages = fs
      .readdirSync(path.join(basePath, dir))
      .filter((packageName) => !/^\./.test(packageName));
    return packages.concat(scopedSuffixPackages.map((p) => path.join(dir, p)));
  } else {
    return packages.concat(dir);
  }
};

//////////////////////////////
// END Helper Methods
//////////////////////////////

const initialBranch = (): ModuleTree =>
  structuredClone({
    name: '',
    version: '',
    dependencies: {},
    isDynamic: false,
    peerDependencies: {},
    peerDependenciesMeta: {},
    resolved: {},
    ref: '',
    moduleLineage: [],
    modules: {},
    modulePath: '',
  });

// This variable will be mutated by the buildModuleTree function
const moduleTreeStore: ModuleTree = initialBranch();

const resolveLineage = (modulePath: string) => {
  const lineage = modulePath
    .split(/\/?node_modules\//)
    .filter((p) => p && p !== ROOT_REF);
  lineage.unshift(ROOT_REF);
  return lineage;
};

const buildModuleTreeRecursive = (
  basePath: string,
  rootBasePath: string,
  currentModuleBranch: ModuleTree = moduleTreeStore,
  isDynamicFlag = false,
): ModuleTree => {
  const packageJsonPath = path.join(basePath, 'package.json');
  const nodeModulesPath = path.join(basePath, 'node_modules');
  const isRootPkg = basePath === rootBasePath;

  if (fs.existsSync(packageJsonPath)) {
    const modulePath = isRootPkg ? '.' : path.relative(rootBasePath, basePath);
    const packageJson = readJsonFile<PackageJson>(packageJsonPath);
    const ref = resolveLineage(modulePath).join(REF_SEPARATOR);
    isDynamicFlag = isDynamicFlag || isDynamicPackage(ref);

    // If we are a nested module, set the moduleTree branch to the current module.
    currentModuleBranch = isRootPkg
      ? currentModuleBranch
      : (currentModuleBranch['modules'][packageJson.name] = initialBranch());

    Object.assign(currentModuleBranch, {
      name: packageJson.name,
      version: packageJson.version,
      isDynamic: isDynamicFlag,
      dependencies: packageJson.dependencies,
      peerDependencies: packageJson.peerDependencies,
      peerDependenciesMeta: packageJson.peerDependenciesMeta,
      ref,
      moduleLineage: resolveLineage(modulePath),
      modules: {},
      modulePath,
    });

    if (fs.existsSync(nodeModulesPath)) {
      const nodeModulesPackages = fs.readdirSync(nodeModulesPath);
      // Filter out hidden directories (eg. .bin)
      // Get fully scoped package names (eg. @dynamic-labs/sdk-api)
      const nodules = nodeModulesPackages
        .filter((dir) => !/^\./.test(dir)) // no hidden directories (eg. .bin)
        .reduce(
          (packages, dir) => mergeScopes(nodeModulesPath, packages, dir),
          [] as string[],
        );

      // Recurse through modules
      nodules.forEach((nmPackage) =>
        buildModuleTreeRecursive(
          path.join(nodeModulesPath, nmPackage),
          rootBasePath,
          currentModuleBranch,
          isDynamicFlag,
        ),
      );
    }
  }

  return currentModuleBranch;
};

export const buildModuleTree = (
  basePath = process.cwd(),
  rootBasePath = process.cwd(),
) => buildModuleTreeRecursive(basePath, rootBasePath);
