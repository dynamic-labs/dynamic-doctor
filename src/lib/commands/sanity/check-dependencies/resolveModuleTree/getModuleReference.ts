import { ModuleTree } from '../types';
import { REF_SEPARATOR, ROOT_REF } from '../constants';

const getModuleReferenceRecursive = (
  moduleTree: ModuleTree,
  ref: string,
  originalRef = ref,
): ModuleTree => {
  // Return the root module tree
  if (ref === ROOT_REF) {
    return moduleTree;
  }

  // Split the ref into module parts
  const refParts = ref.split(REF_SEPARATOR);

  // Grab the left most module
  let nextModule = refParts.shift();
  if (nextModule === ROOT_REF) {
    nextModule = refParts.shift();
  }

  if (nextModule) {
    // As we are descending the tree, we need to make sure the next module exists
    if (!moduleTree.modules[nextModule]) {
      throw new Error(
        `Invalid ref: ${originalRef}, could not find next module: ${nextModule}`,
      );
    }

    // If there are no more parts, we have found the module
    if (refParts.length === 0) {
      return moduleTree.modules[nextModule];
    } else {
      // Otherwise, recurse down the tree
      return getModuleReferenceRecursive(
        moduleTree.modules[nextModule],
        refParts.join(REF_SEPARATOR),
        originalRef,
      );
    }
  }

  // If we have not returned by now, the ref is invalid
  throw new Error(`Invalid ref: ${originalRef}`);
};

/**
 * Get the module reference in the tree given a ref string
 *
 * This does not resolve for missing modules, the module MUST exist in the tree
 * @example getModuleReference('module1:module2:module3', moduleTree)
 * @param ref A colon separated string of module names to traverse
 * @param moduleTree The root module tree
 * @returns The module tree at the ref
 */
export const getModuleReference = (
  moduleTree: ModuleTree,
  ref: string,
): ModuleTree => getModuleReferenceRecursive(moduleTree, ref);
