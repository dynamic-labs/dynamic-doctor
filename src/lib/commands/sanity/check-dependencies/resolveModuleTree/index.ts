//export * from './resolveModuleTree';
export * from './getModuleReference';
export * from './resolveModuleTree';
