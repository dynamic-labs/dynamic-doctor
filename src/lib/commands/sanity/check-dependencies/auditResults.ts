import { REF_SEPARATOR } from './constants';
import { getModuleReference } from './resolveModuleTree';
import { AuditResults, ModuleTree } from './types';

const filterDynamicModules = (moduleTree: ModuleTree) => {
  return Object.fromEntries(
    Object.entries(moduleTree.modules).filter(([, module]) => module.isDynamic),
  );
};

const auditResultsRecursive = (
  moduleTree: ModuleTree,
  auditResultsStore: AuditResults = {},
) => {
  const dynamicModuleTree = filterDynamicModules(moduleTree);
  Object.values(dynamicModuleTree).forEach((module) => {
    Object.entries(module.resolved).forEach(
      ([resolvedModuleName, resolvedModule]) => {
        if (!resolvedModule.satisfied && !resolvedModule.optional) {
          const key = `${module.ref}${REF_SEPARATOR}${resolvedModuleName}`;
          auditResultsStore[key] = {
            ...resolvedModule,
            name: resolvedModuleName,
          };
          if (resolvedModule.ref) {
            const referencedModule = getModuleReference(
              moduleTree,
              resolvedModule.ref,
            );
            auditResultsStore[key].dependency = referencedModule;
            auditResultsRecursive(referencedModule, auditResultsStore);
          }
        }
      },
    );
  });

  return auditResultsStore;
};

export const auditResults = (moduleTree: ModuleTree) => {
  return auditResultsRecursive(moduleTree);
};
