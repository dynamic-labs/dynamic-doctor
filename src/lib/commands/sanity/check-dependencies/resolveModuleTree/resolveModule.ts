import { ModuleTree } from '../types';
import { REF_SEPARATOR } from '../constants';
import { getModuleReference } from './getModuleReference';

const getNextRefStep = (ref: string): string => {
  const modules = ref.split(REF_SEPARATOR);
  modules.pop(); // Drop last element
  const nextRef = modules.join(REF_SEPARATOR);
  return nextRef;
};

const resolveModuleRecursive = (
  moduleTree: ModuleTree,
  dependencyRef: string,
  moduleName: string,
  rootModuleTree: ModuleTree = moduleTree,
): string | null => {
  const currentModuleTree = getModuleReference(rootModuleTree, dependencyRef);
  if (currentModuleTree.modules[moduleName]) {
    return `${dependencyRef}${REF_SEPARATOR}${moduleName}`;
  } else if (dependencyRef === '.') {
    return null;
  }

  const nextRef = getNextRefStep(dependencyRef);
  const nextModuleTree = getModuleReference(rootModuleTree, nextRef);

  return resolveModuleRecursive(
    nextModuleTree,
    nextRef,
    moduleName,
    rootModuleTree,
  );
};

export const resolveModule = (
  moduleTree: ModuleTree,
  startingRef: string,
  moduleName: string,
) => resolveModuleRecursive(moduleTree, startingRef, moduleName);
