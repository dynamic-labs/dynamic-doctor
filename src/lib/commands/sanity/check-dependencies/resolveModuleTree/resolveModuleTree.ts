import { ModuleTree } from '../types';
import { REF_SEPARATOR } from '../constants';
import { getModuleReference } from './getModuleReference';
import { resolveModule } from './resolveModule';
import satisfies from 'semver/functions/satisfies';

const resolveModuleTreeModule = (
  type: 'dependency' | 'peerDependency',
  moduleName: string,
  module: ModuleTree,
  moduleTree: ModuleTree,
) => {
  const depType = type === 'dependency' ? 'dependencies' : 'peerDependencies';
  const deps = moduleTree[depType];

  if (deps && deps[moduleName]) {
    moduleTree.resolved[moduleName] = {
      type,
      ref: `${moduleTree.ref}${REF_SEPARATOR}${moduleName}`,
      optional: false,
      satisfied: satisfies(module.version, deps[moduleName]),
      requiredVersion: deps[moduleName],
      installedVersion: module.version,
    };
  }
};

const resolveModuleTreeModules = (
  moduleTree: ModuleTree,
  rootModuleTree: ModuleTree = moduleTree,
) => {
  Object.entries(moduleTree.modules).forEach(([moduleName, module]) => {
    resolveModuleTreeModule('dependency', moduleName, module, moduleTree);
    resolveModuleTreeModule('peerDependency', moduleName, module, moduleTree);
    resolveModuleTreeRecursive(module, rootModuleTree);
  });
};

const resolveModuleTreeDependenciesModule = (
  type: 'dependency' | 'peerDependency',
  moduleTree: ModuleTree,
  rootModuleTree: ModuleTree,
) => {
  const depType = type === 'dependency' ? 'dependencies' : 'peerDependencies';
  const deps = moduleTree[depType];

  if (deps) {
    Object.entries(deps).forEach(([moduleName, version]) => {
      // Handle npm aliases
      const aliasMatch = version.match(/^npm:(.*)@(.*)/);
      if (aliasMatch) {
        version = aliasMatch[2];
      }

      if (!moduleTree.resolved[moduleName]) {
        const resolvedRef = resolveModule(
          rootModuleTree,
          moduleTree.ref,
          moduleName,
        );

        const optional =
          moduleTree.peerDependenciesMeta?.[moduleName]?.optional ?? false;

        moduleTree.resolved[moduleName] = {
          type,
          ref: resolvedRef,
          optional,
          satisfied: false,
          requiredVersion: version,
          installedVersion: null,
        };

        if (resolvedRef) {
          const module = getModuleReference(rootModuleTree, resolvedRef);
          moduleTree.resolved[moduleName].satisfied = satisfies(
            module.version,
            version,
          );
          moduleTree.resolved[moduleName].installedVersion = module.version;
          module.isDynamic = moduleTree.isDynamic;
          resolveModuleTreeRecursive(module, rootModuleTree);
        }
      }
    });
  }
};

const resolveModuleTreeDependencies = (
  moduleTree: ModuleTree,
  rootModuleTree: ModuleTree = moduleTree,
) => {
  resolveModuleTreeDependenciesModule('dependency', moduleTree, rootModuleTree);
  resolveModuleTreeDependenciesModule(
    'peerDependency',
    moduleTree,
    rootModuleTree,
  );
};

const resolveModuleTreeRecursive = (
  moduleTree: ModuleTree,
  rootModuleTree: ModuleTree = moduleTree,
) => {
  resolveModuleTreeModules(moduleTree, rootModuleTree);
  resolveModuleTreeDependencies(moduleTree, rootModuleTree);
  return moduleTree;
};

export const resolveModuleTree = (moduleTree: ModuleTree) => {
  moduleTree = structuredClone(moduleTree);
  return resolveModuleTreeRecursive(moduleTree);
};
