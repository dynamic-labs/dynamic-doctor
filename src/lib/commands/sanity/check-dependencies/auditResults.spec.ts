import { buildMocks, fixtures } from './__fixtures__/fixtures';
import { auditResults } from './auditResults';
import { buildModuleTree } from './buildModuleTree';
import { resolveModuleTree } from './resolveModuleTree';

jest.mock('fs');

describe('auditResults', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  describe.each(Object.entries(fixtures))('fixture %s', (key, value) => {
    const basePath = '/specs';
    beforeEach(() => {
      buildMocks(value, basePath);
    });
    it('should audit a module tree', () => {
      const moduleTree = buildModuleTree(basePath, basePath);
      const resolvedModuleTree = resolveModuleTree(moduleTree);
      const audit = auditResults(resolvedModuleTree);
      expect(audit).toMatchSnapshot();
    });
  });
});
