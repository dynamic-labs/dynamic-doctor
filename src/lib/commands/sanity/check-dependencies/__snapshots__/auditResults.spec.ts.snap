// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auditResults fixture fixture_01 should audit a module tree 1`] = `{}`;

exports[`auditResults fixture fixture_02 should audit a module tree 1`] = `
{
  ".:@dynamic-labs/example:exampleDepA": {
    "installedVersion": null,
    "name": "exampleDepA",
    "optional": false,
    "ref": null,
    "requiredVersion": "1.70.1",
    "satisfied": false,
    "type": "dependency",
  },
  ".:@dynamic-labs/example:exampleDepB": {
    "installedVersion": null,
    "name": "exampleDepB",
    "optional": false,
    "ref": null,
    "requiredVersion": "^5.25.0",
    "satisfied": false,
    "type": "dependency",
  },
  ".:@dynamic-labs/example:exampleDepC": {
    "installedVersion": null,
    "name": "exampleDepC",
    "optional": false,
    "ref": null,
    "requiredVersion": "1.1.0-alpha.10",
    "satisfied": false,
    "type": "dependency",
  },
  ".:@dynamic-labs/example:examplePeerA": {
    "installedVersion": null,
    "name": "examplePeerA",
    "optional": false,
    "ref": null,
    "requiredVersion": "^1.19.13",
    "satisfied": false,
    "type": "peerDependency",
  },
}
`;

exports[`auditResults fixture fixture_03 should audit a module tree 1`] = `
{
  ".:@dynamic-labs/ethers-v5:ethers": {
    "installedVersion": null,
    "name": "ethers",
    "optional": false,
    "ref": null,
    "requiredVersion": "^5",
    "satisfied": false,
    "type": "peerDependency",
  },
  ".:@dynamic-labs/ethers-v6:ethers": {
    "installedVersion": null,
    "name": "ethers",
    "optional": false,
    "ref": null,
    "requiredVersion": "^6",
    "satisfied": false,
    "type": "peerDependency",
  },
  ".:@dynamic-labs/wagmi-connector:eventemitter3": {
    "dependency": {
      "dependencies": undefined,
      "isDynamic": true,
      "moduleLineage": [
        ".",
        "eventemitter3",
      ],
      "modulePath": "node_modules/eventemitter3",
      "modules": {},
      "name": "eventemitter3",
      "peerDependencies": undefined,
      "peerDependenciesMeta": undefined,
      "ref": ".:eventemitter3",
      "resolved": {},
      "version": "4.0.7",
    },
    "installedVersion": "4.0.7",
    "name": "eventemitter3",
    "optional": false,
    "ref": ".:eventemitter3",
    "requiredVersion": "5.0.1",
    "satisfied": false,
    "type": "peerDependency",
  },
  ".:@dynamic-labs/wallet-connector-core:eventemitter3": {
    "dependency": {
      "dependencies": undefined,
      "isDynamic": true,
      "moduleLineage": [
        ".",
        "eventemitter3",
      ],
      "modulePath": "node_modules/eventemitter3",
      "modules": {},
      "name": "eventemitter3",
      "peerDependencies": undefined,
      "peerDependenciesMeta": undefined,
      "ref": ".:eventemitter3",
      "resolved": {},
      "version": "4.0.7",
    },
    "installedVersion": "4.0.7",
    "name": "eventemitter3",
    "optional": false,
    "ref": ".:eventemitter3",
    "requiredVersion": "5.0.1",
    "satisfied": false,
    "type": "peerDependency",
  },
  ".:@perawallet/connect:algosdk": {
    "installedVersion": null,
    "name": "algosdk",
    "optional": false,
    "ref": null,
    "requiredVersion": "^2.1.0",
    "satisfied": false,
    "type": "peerDependency",
  },
}
`;
